src/config/account_config.rs

```diff
@@ -190,8 +190,14 @@ impl<'a> AccountConfig {
                 imap_passwd_cmd: config.imap_passwd_cmd.clone(),
             }),
             DeserializedAccountConfig::Maildir(config) => {
+                let maildir_path_str = config
+                    .maildir_dir
+                    .to_string_lossy();
+                let expanded_path = shellexpand::full(&maildir_path_str)?
```

Comment:
Just a side note on how the config folder is now organized. You have 3 main files:

- `deserialized_config` which contains the deserialized config file by serde (you can see it as the raw translation of the TOML file)
- `deserialized_account_config` which contains the deserialized struct of an account (from the HashMap located in `deserialized_config`)
- `account_config` which contains the final config of the currently selected account (there is no more config, it has been merged with the account for clarity)

---

`shellexpand::full` expects a `AsRef<str>`, but we have a `PathBuf`. I think the simplest is to change the raw deserialized type of `maildir_dir` (in `deserialized_account_config`, last line) from `PathBuf` to `String`. This way you could simplify all what you did by:

```rust
BackendConfig::Maildir(MaildirBackendConfig {
    maildir_dir: shellexpand::full(&config.maildir_dir)?.to_string().into(),
})
```

Reply:
I get the following if I use your suggestion:
```rust
error[E0277]: the trait bound `PathBuf: AsRef<str>` is not satisfied
   --> src/config/account_config.rs:194:52
    |
194 |                     maildir_dir: shellexpand::full(&config.maildir_dir)?.to_string().into(),
    |                                  ----------------- ^^^^^^^^^^^^^^^^^^^ the trait `AsRef<str>` is not implemented for `PathBuf`
    |                                  |
    |                                  required by a bound introduced by this call
    |
    = help: the following implementations were found:
              <PathBuf as AsRef<OsStr>>
              <PathBuf as AsRef<Path>>
note: required by a bound in `full`
   --> /home/tornax/.cargo/registry/src/github.com-1ecc6299db9ec823/shellexpand-2.1.0/src/lib.rs:303:9
    |
303 |     SI: AsRef<str>,
    |         ^^^^^^^^^^ required by this bound in `full`
```

Reply:
Because you did not change the type of deserialized `maildir_dir` property (check the `config/deserialized_account_config.rs` file, last line).

Reply:
Sorry, I wasn't really aware of your sentence here:
> I think the simplest is to change the raw deserialized type of maildir_dir (in deserialized_account_config, last line) from PathBuf to String.

It's working now. Hm... But I'm wondering if it isn't better if we are still using `Path`, because otherwise there might be some restrictions for some people for their file-/directorynames.

Reply:
By the way, thank you for the eplanation of these 3 files :)

Reply:
> But I'm wondering if it isn't better if we are still using `Path`, because otherwise there might be some restrictions for some people for their file-/directorynames

I think it does not matter. You only changed the type of the deserialized config, but the type of the final config remains `PathBuf` (note the `.into()` at the end of your changes).

Reply:
oh yes, you're rigth :+1: